### PR TITLE
Multiple dist dl sampler

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

This PR aims to enable the `custom pool` could be shared/used by any number of `DistDataLoaders`. This feature is required when try to inference during train. see more details in https://github.com/dmlc/dgl/issues/4235.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tests/distributed/test_mp_dataloader.py [hunk 1]
```diff
@@ -1,19 +1,24 @@
-import dgl
```
Comment (Rhett-Ying) [inline tests/distributed/test_mp_dataloader.py:1]: I refined/formatted this file, pls pay attention to the newly added `test_multiple_dist_dataloaders` which is at the end of file. this should be fine.

File: python/dgl/distributed/dist_context.py [hunk 2]
```diff
@@ -122,8 +123,11 @@ def __init__(self, num_workers, rpc_config):
         """
         ctx = mp.get_context("spawn")
         self.num_workers = num_workers
-        self.queue_size = num_workers * 4
+        # As pool could be used by any number of dataloaders, queues
+        # should be able to take infinite elements to avoid dead lock.
+        self.queue_size = 0
         self.result_queue = ctx.Queue(self.queue_size)
+        self.results = {}
```
Comment (jermainewang) [inline python/dgl/distributed/dist_context.py:130]: Add some comments about the key and values

File: python/dgl/distributed/dist_context.py [hunk 3]
```diff
@@ -167,16 +172,20 @@ def submit_task_to_all_workers(self, func, args):
 
     def get_result(self, dataloader_name, timeout=1800):
         """Get result from result queue"""
-        result_dataloader_name, result = self.result_queue.get(timeout=timeout)
-        assert result_dataloader_name == dataloader_name
-        return result
+        if dataloader_name not in self.results:
+            raise DGLError(f"{dataloader_name} not found.")
```
Comment (jermainewang) [inline python/dgl/distributed/dist_context.py:176]:
```suggestion
            raise DGLError(f"Got result from an unknown dataloader {dataloader_name}.")
```

File: python/dgl/distributed/dist_context.py [hunk 4]
```diff
@@ -167,16 +172,20 @@ def submit_task_to_all_workers(self, func, args):
...
+        while len(self.results[dataloader_name]) == 0:
+            dl_name, data = self.result_queue.get(timeout=timeout)
```
Comment (jermainewang) [inline python/dgl/distributed/dist_context.py:178]: What will happen if `self.result_queue.get` timed out? Are we able to catch that?

**Review by jermainewang — APPROVED** [review]
Generally LGTM. Since the original task is to enable model inference during training, is it possible to provide an example?

File: python/dgl/distributed/dist_context.py (hunk 4 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_context.py:178]: raises the [queue.Empty](https://docs.python.org/3/library/queue.html#queue.Empty) exception if no item was available within that time.
As this should never happen in normal runnings, we don't catch such exception and just let it abort the app. I think this behavior is acceptable. If it happens, something went wrong in the sampler processes.

File: python/dgl/distributed/dist_context.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_context.py:130]: added.

File: python/dgl/distributed/dist_context.py (hunk 3 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_context.py:176]: accepted.

**Comment by Rhett-Ying** [comment]
> Generally LGTM. Since the original task is to enable model inference during training, is it possible to provide an example?

such example is more about end2end test instead of unit test I added here. For now, we don't have such end2end test framework or environment for the moment. But I believe the unit tests I added simulate the key part of end2end scenario. `iterating another dataloader during iterating on one dataloader` is supposed to mimic `do inference during train`, I think.

**Comment by jermainewang** [comment]
> > Generally LGTM. Since the original task is to enable model inference during training, is it possible to provide an example?
> 
> such example is more about end2end test instead of unit test I added here. For now, we don't have such end2end test framework or environment for the moment. But I believe the unit tests I added simulate the key part of end2end scenario. `iterating another dataloader during iterating on one dataloader` is supposed to mimic `do inference during train`, I think.

Sure. Could you create a task about the end2end test for tracking purposes?

**Pull request merged**